File: benchmarks/compare/raptor/go/timeparser.go
package raptor

import (
	"strconv"
	"strings"
)

// TimeParser parses "HH:MM:SS" strings into seconds from midnight, caching
// results.
type TimeParser struct {
	cache map[string]Time
}

// NewTimeParser constructs a TimeParser with an empty cache.
func NewTimeParser() *TimeParser {
	return &TimeParser{cache: map[string]Time{}}
}

// GetTime converts a time string to seconds from midnight.
func (p *TimeParser) GetTime(time string) Time {
	if v, ok := p.cache[time]; ok {
		return v
	}

	parts := strings.Split(time, ":")
	hh, _ := strconv.ParseInt(parts[0], 10, 64)
	mm, _ := strconv.ParseInt(parts[1], 10, 64)
	ss, _ := strconv.ParseInt(parts[2], 10, 64)

	v := hh*60*60 + mm*60 + ss
	p.cache[time] = v
	return v
}

File: benchmarks/compare/raptor/rust/src/time_parser.rs
//! "HH:MM:SS" → seconds-from-midnight, cached (port of src/gtfs/TimeParser.ts).

use std::collections::HashMap;

use crate::gtfs::Time;

#[derive(Debug, Default)]
pub struct TimeParser {
    cache: HashMap<String, Time>,
}

impl TimeParser {
    pub fn new() -> Self {
        TimeParser::default()
    }

    pub fn get_time(&mut self, time: &str) -> Time {
        if let Some(&v) = self.cache.get(time) {
            return v;
        }

        let parts: Vec<&str> = time.split(':').collect();
        let hh: i64 = parts.first().map_or(0, |s| s.parse().unwrap_or(0));
        let mm: i64 = parts.get(1).map_or(0, |s| s.parse().unwrap_or(0));
        let ss: i64 = parts.get(2).map_or(0, |s| s.parse().unwrap_or(0));

        let seconds = hh * 60 * 60 + mm * 60 + ss;
        self.cache.insert(time.to_string(), seconds);

        seconds
    }
}
